File: sorting/shell/shell.js
"use strict";

function shell(arr) {
    let gap = Math.floor(arr.length / 2);
    while (gap >= 1) {

        for (let i = 0; i < arr.length - gap; i++) {
            let pos = i + gap;
            while (pos - gap >= 0 && arr[pos] < arr[pos - gap]) {
                // This doesn't seem to work yet despite destructuring set
                //[arr[pos], arr[pos - gap]] = [arr[pos - gap], arr[pos]];
                const temp = arr[pos];
                arr[pos] = arr[pos - gap];
                arr[pos - gap] = temp;
                pos -= gap;
            }
        }
        gap = Math.floor(gap / 2);
    }
}

if (require.main === module) {
    console.log("Shell sort");
    let arr = [8, 9, 7, 5, 2, 4, 6, 1, 3];
    shell(arr);
    console.log(arr);
}

File: sorting/shell/shell.py
def shell(arr):
    gap = len(arr) // 2
    while gap >= 1:
        for i in xrange(len(arr)):
            if i + gap > len(arr) - 1:
                break
            insertion_sort_gap(arr, i, gap)
        gap //= 2


def insertion_sort_gap(arr, start, gap):
    pos = start + gap
    while pos - gap >= 0 and arr[pos] < arr[pos - gap]:
        arr[pos], arr[pos - gap] = arr[pos - gap], arr[pos]
        pos -= gap


if __name__ == '__main__':
    arrs = [
        [4, 3, 3, 7, 6, -1, 10, 3, 8, 4],
        [10, 11, 9, 8, 13, 21],
        [99, 87, 76, 65, 54, 43, 32, 21, 10, -1],
    ]
    for arr in arrs:
        print("unsorted: {}".format(arr))
        shell(arr)
        print("sorted: {}".format(arr))
